evaluator.go

```diff
@@ -0,0 +1,278 @@
+package CloudForest
+
+import (
+	"fmt"
+	"strconv"
+)
+
+const leafFeature = -1
+
+// the evaluator interface implements high performance
+// decision tree evaluation strategies.
+// The PiecewiseFlatForest and the ContiguousFlatForest
+// provide faster analogs of the Predict function
+type NumEvaluator interface {
+	EvaluateNum(fm *FeatureMatrix) []float64
+}
+
+type CatEvaluator interface {
+	EvaluateCat(fm *FeatureMatrix) []string
+}
+
+type FlatNode struct {
+	Feature   int    `json:"feature"`
+	Value     string `json:"value"`
+	LeftChild uint32 `json:"leftchild"`
+}
+
+type FlatTree struct {
+	Nodes []*FlatNode `json:"nodes"`
+}
+
+func NewFlatTree(root *Node) *FlatTree {
+	f := &FlatTree{make([]*FlatNode, 1)}
+	f.recurse(root, 0)
+	return f
+}
+
+func (f *FlatTree) recurse(n *Node, idx uint32) {
+	if n.Left == nil && n.Right == nil {
+		f.Nodes[idx] = &FlatNode{
+			Feature: leafFeature,
+			Value:   n.Pred,
+		}
+		return
+	}
+	leftChild := uint32(len(f.Nodes))
+	f.Nodes = append(f.Nodes, make([]*FlatNode, 2)...)
+	var value string
+	switch x := n.CodedSplit.(type) {
+	case float64, int:
+		value = fmt.Sprintf("%v", x)
```

Comment:
Why do you convert this to a string, and then back to a float later? Seems better to just keep the raw values